src/librbd/ManagedLock.h

```diff
@@ -0,0 +1,184 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_MANAGED_LOCK_H
+#define CEPH_LIBRBD_MANAGED_LOCK_H
+
+#include "include/int_types.h"
+#include "include/Context.h"
+#include "include/rados/librados.hpp"
+#include "cls/lock/cls_lock_types.h"
+#include "common/Mutex.h"
+#include <list>
+#include <string>
+#include <utility>
+
+class ContextWQ;
+
+namespace librbd {
+
+template <typename Watcher>
```

Comment:
I tried to change the template to ImageCtx, and use the Watcher base class but it required so many changes to the unit tests that I don't think it's worth it. The main problem is that by using the Watcher base class without template it, mocking the watcher in very hard, as it required mocking using inheritance.